### ISSUE TITLE
UI is not Accessible (screen reader support)

### ISSUE THREAD

**Issue by webprofusion-chrisc** [webprofusion/certify#272]
Our UI is currently not optimised in any way for Accessibility and has no specific screen reader support. 

This issue is to track accessibility progress and to prompt volunteers to submit their first PR! If you do want to contribute changes please split PRs into per-component or area to avoid having one big PR to merge as the 4.x UI is very much still under development.

**Comment by Cambridgeport90** [comment]
@webprofusion-chrisc, thanks for the hand with this. What I can tell you. The tabs are very nicely labeled, and I can read their titles regardless of which screen reader is in use. It's the buttons on each tab that don't seem to be labeled; screen readers will say "button" instead of reading the actual label. Normally means that the accessible description field is missing from the control, though without much knowledge (my coding skills are too weak to be able to recognize it), of whether you're using custom controls or not, those might have to be added. The other things I notice is that the "managed sites" tab seems to have the "add" button out of the tab order; screen readers can read it when you run a full screen review command, but you can't tab to it, and then there is this weird unlabeled edit field which is simply identified as "edit" when you tab to it. If you run NVDA on your test machine, you'll be able to see what I'm talking about. Their site is [](https://nvda-project.org)

**Comment by webprofusion-chrisc** [comment]
One important consideration here is that Microsoft made significant changes for .net 4.7.1, https://blogs.msdn.microsoft.com/dotnet/2017/09/21/net-framework-4-7-1-accessibility-and-wpf-improvements/ so if we don't target that then it looks like we'd be in the 'legacy' category. However it's unclear what level of support the installed base of screen readers have for this version, so it may be safest not to target the new features? If we did target 4.7.1 we'd also need to either force everyone not already on that version to upgrade or have two builds.

**Comment by Cambridgeport90** [comment]
Oh boy. I never even thought of that. Most blind folks I know are on Windows 10, which come with 4.7 by default, so not sure screen readers of any kind care which version it is. I’ll read through that though, and see if I can find anything. And I’ll talk to some people I know as well in my own group that also do .NET stuff.

**Comment by webprofusion-chrisc** [comment]
@Cambridgeport90 thanks for the feedback, is NVDA one of the most popular tools? Does anyone use the builtin windows Narrator or is that really good enough for the job? The app does use some custom controls http://mahapps.com/ but I'm sure we just need to be specific about setting accessible properties.

**Comment by Cambridgeport90** [comment]
Narrator does work, though I’ve tried it more with built-in applications rather than third-party ones. NVDA is open source as well, so that’s why I always recommend developers including testing with it in their workflows. But I’ll be sure to run you some more testing with Narrator by the end of the week.

**Comment by webprofusion-chrisc** [comment]
@Cambridgeport90  thanks, I wouldn't worry too much about analysing the current release of the app (which is 3.0.11)  as the new version in the works has hundreds of UI changes. Once I can get a test release out (hopefully next week, maybe the week after) I'll update here to let people know to give it a try. The next version has large bodies of instructional text which weren't there before as well as a very different tab layout in general.

**Comment by Cambridgeport90** [comment]
Awesome. And I would encourage you to consider sticking with WPF, considering it can be made accessible. It’s all about including the appropriate properties.

**Comment by Cambridgeport90** [comment]
I just downloaded and ran testing on the latest version of the UI for accessibility. You've definitely still got some work to do. I can contact some of the folks from Microsoft I'm in contact with if you need more support on the accessibility documentation, but a huge issue I see right now has to do with when creating a new certificate, for instance, the button (in that case) is labeled; others just read as "button" (screen reader used doesn't matter), and under the combobox where you can select which domain to use with the certificate, you get from any screen reader used "certify.models.bindings", instead of just the domain you're focused on being read. You might ought to check the UIA description for that control; sounds like the control doesn't have a label. Also, in that same location, there are edit fields that also don't have accessible labels; the one for the cert display name as well as the one for the domain, if you want to add it manually.  

Needless to say, this causes a blind person trying to use this to have to go through some unnecessary trial and error to tap the correct field when that doesn't have to be the case. Note that this is with version 4, I see these issues. I also see this other control labeled "thumb control". I have no idea what it's supposed to do or what the function is. It shows up next to the tabs, regardless of which of them you're focused on. Adding descriptions for controls is built right into the framework no matter which one you're targeting; I remember you had previously expressed that as a concern. Your simplest bet to fix these issues is to go through Microsoft's documentation. I can link you to any specific one you need as well. Even easier, take a stab at downloading NVDA and viewing the interface as I do; that will show you much better than I ever could.

**Comment by webprofusion-chrisc** [comment]
@Cambridgeport90 thanks for your feedback, I onlt just noticed your reply. There hasn't been much work on accessibility for v4 yet, the main focus so far has been on stabilising features and larger bugs but I do plan to do a couple of accessibility passes. As you say, this is mostly fixing control labels (AutomationProperties Name in WPF) but in some cases things need reorganised. The aim is to have the core UI covered for the final v4.0 release. The delay is really just due to time constraints, so if you know anyone that has the time and ability to directly submit fixes that would be great, otherwise we will get there eventually either way.

**Comment by Cambridgeport90** [comment]
I figured as much. My comment was more to just give you a quick progress update. I figured it hadn’t fallen by the wayside. I’ll check with one of the folks in my group; I do know a couple of .NET coders. But thanks a million for checking in!

**Comment by webprofusion-chrisc** [comment]
Addressed in Certify management Hub.
